src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PENamespaceSymbol.cs

```diff
@@ -42,6 +42,14 @@ internal abstract class PENamespaceSymbol
         /// </summary>
         private Dictionary<string, TypeDefinitionHandle> _lazyNoPiaLocalTypes;
 
+        /// <summary>
+        /// Caches the last call to GetMembers(string) when that member is a namespace. Multiple types
+        /// within the same namespace are often requested in sequence, so this allows the second and
+        /// subsequent request for members with the same name to share a result without looking up the
+        /// values or allocating.
+        /// </summary>
+        private ImmutableArray<Symbol> _lastGetMembersResult;
```

Comment:
>_lastGetMembersResult [](http://example.com/codeflow?start=39&length=21)

What exactly is the use case scenario? It sounds like this is something that a consumer can do instead.

Reply:
This occurs when a user requests two symbols such as:

* ``System.Collections.Generic.IList`1``
* ``System.Collections.Generic.List`1``

With the new optimization, the named lookups for `System`, `Collections`, and `Generic` will all be direct cache hits.

Reply:
> Performance traces suggest that this is a hot path

For what specific scenarios? The PR description has zero information, not to mention perf comparisons.

Reply:
@AlekseyTs Yes, sorry for the missing info.

This PR was opened per a discussion with @sharwell where he rooted down some allocations in a performance trace I had to this code path. It was purely his investigations and observations that this change would improve those allocations. I only implemented his suggestion.

The idea, per my understanding, is as mentioned in the comment. A quick demonstration is the example by @sharwell:

Consider the following calls to `GetTypeByMetadataName`:

```csharp
compilation.GetTypeByMetadataName("System.Collections.Generic.IList`1");
compilation.GetTypeByMetadataName("System.Collections.Generic.List`1");
```

So, for the first call, we start off with `PENamespaceSymbol` representing `System`, and we call `GetMembers("Collections")`, then on the result of that we call `GetMembers("Generic")` then we find `IList<T>`

Then, most of that work is repeated again when we request `List<T>`.

By caching the last lookup in `System` namespace, we reduce the work when `Collections` will be queried again. This goes recursively to nested namespaces as well.

Reply:
@AlekseyTs I have updated this doc comment. Let me know if it's more clear now?